enclave/crypto/cert.c

```diff
@@ -737,17 +741,27 @@ oe_result_t oe_cert_chain_read_pem(
     if (!pem_data || !pem_size || !chain)
         OE_RAISE(OE_INVALID_PARAMETER);
 
-    /* Must have pem_size-1 non-zero characters followed by zero-terminator */
-    if (strnlen((const char*)pem_data, pem_size) != pem_size - 1)
-        OE_RAISE(OE_INVALID_PARAMETER);
+    // mbedtls_x509_crt_parse() requires a trailing zero in its input buffer.
+    // If the input pem_data buffer does not have a trailing zero,
+    // we allocate a tmp buffer to add it.
```

Comment:
If I understand correctly, this PR tries to handle those no-trailing-null pem-formated cert chain gracefully (by adding one), instead of just throwing OE_INVALID_PARAMETER? 

As far as I know, correct me if am wrong, the pem-formated cert chain returned by Intel DCAP client and 1.2+ version Azure DCAP client does have the trailing null character. I am curious in which case you can trigger this bug?

Reply:
Yes your understanding is accurate.

The test platform that I use is ICL notebook running Ubuntu 18.04 (not an Azure VM). It is configured with libsgx-dcap-ql v1.5, the DCAP reference QPL (DCAP client), and the reference PCK cert caching service sgx-dcap-pccs v1.5

In this setup, the pem-formatted cert chain has no trailing null, and the existing implementation fails (both the host and enclave sides).

This PR has no effect if the buffer does have a trailing null.

Reply:
Thank you very much for clarifying. It is great that we have the fault tolerance capability by this PR. If I understand correctly, based on the contract between qpl/ql and OE, it is qpl's responsibility to properly add trailing null character. Is that correct?